### PR TITLE
frame-system: Add `last_runtime_upgrade_spec_version`

### PR DESCRIPTION
**Pull request by bkchr**
Adds a function for querying the last runtime upgrade spec version. This can be useful for when writing runtime level migrations to ensure that they are not executed multiple times. An example would be a session key migration.

### CODE REVIEW

File: substrate/frame/executive/src/lib.rs [hunk 1]
```diff
@@ -487,6 +487,12 @@ where
 		let mut weight = Weight::zero();
 		if Self::runtime_upgraded() {
 			weight = weight.saturating_add(Self::execute_on_runtime_upgrade());
+
+			frame_system::LastRuntimeUpgrade::<System>::put(
+				frame_system::LastRuntimeUpgradeInfo::from(
+					<System::Version as frame_support::traits::Get<_>>::get(),
+				),
+			);
```
Comment (liamaharon) [inline substrate/frame/executive/src/lib.rs:495]: How was this value being set prior to this PR?

File: substrate/frame/executive/src/lib.rs [hunk 2]
```diff
@@ -503,19 +509,14 @@ where
 		frame_system::Pallet::<System>::note_finished_initialize();
 	}
 
-	/// Returns if the runtime was upgraded since the last time this function was called.
+	/// Returns if the runtime was upgraded.
+	///
+	/// based on [`frame_system::LastRuntimeUpgrade`]
```
Comment (liamaharon) [inline substrate/frame/executive/src/lib.rs:514]:
```suggestion
	/// Returns if the runtime has been upgraded, using [`frame_system::LastRuntimeUpgrade`].
```

Comment (ggwpez) [reply substrate/frame/executive/src/lib.rs:514]: `if` should probably be `whether`.

File: substrate/frame/executive/src/lib.rs (hunk 1 above)
Comment (ggwpez) [reply substrate/frame/executive/src/lib.rs:495]: In the `runtime_upgraded`.

File: substrate/frame/executive/src/lib.rs [hunk 3]
```diff
@@ -1519,6 +1514,11 @@ mod tests {
 
 	#[test]
 	fn all_weights_are_recorded_correctly() {
+		// Reset to get the correct new genesis below.
+		RuntimeVersionTestValues::mutate(|v| {
+			*v = sp_version::RuntimeVersion { spec_version: 0, ..Default::default() }
+		});
```
Comment (ggwpez) [inline substrate/frame/executive/src/lib.rs:1520]:
```suggestion
		RuntimeVersionTestValues::take();
```

File: substrate/frame/executive/src/lib.rs [hunk 4]
```diff
@@ -1535,6 +1535,12 @@ mod tests {
 				Digest::default(),
 			));
 
+			// Reset the last runtime upgrade info, to make the second call to `on_runtime_upgrade`
+			// succeed.
+			LastRuntimeUpgrade::<Runtime>::set(Some(
+				sp_version::RuntimeVersion { spec_version: 0, ..Default::default() }.into(),
+			));
```
Comment (ggwpez) [inline substrate/frame/executive/src/lib.rs:1542]:
```suggestion
			LastRuntimeUpgrade::<Runtime>::take();
```

**Review by ggwpez — APPROVED** [review]

**Review by liamaharon — APPROVED** [review]

**Review by gpestana — APPROVED** [review]

**Comment by yjhmelody** [comment]
Wonder if `docify` could be a dev-dep?

**Pull request merged**